Isabelle fails to build image with HOL-Library

**To Reproduce**
Steps to reproduce the behavior:
1. `isabelle jedit -l "HOL-Library"`
2. An `Isabelle build` window will appear. Wait for the build to finish.
3. Observe the following error in the build log:

<details>

```
Build started for Isabelle/HOL-Library ...
Building HOL-Library ...
HOL-Library: theory HOL-Library.AList
HOL-Library: theory HOL-Library.Adhoc_Overloading
HOL-Library: theory HOL-Library.Monad_Syntax
HOL-Library: theory HOL-Library.State_Monad
HOL-Library: theory HOL-Library.BNF_Axiomatization
HOL-Library: theory HOL-Library.BNF_Corec
HOL-Library: theory HOL-Library.DAList
HOL-Library: theory HOL-Library.Boolean_Algebra
HOL-Library: theory HOL-Library.Cancellation
HOL-Library: theory HOL-Library.Multiset
HOL-Library: theory HOL-Library.Case_Converter
HOL-Library: theory HOL-Library.Code_Lazy
HOL-Library: theory HOL-Library.Simps_Case_Conv
HOL-Library: theory HOL-Library.Extended
HOL-Library: theory HOL-Library.DAList_Multiset
HOL-Library: theory HOL-Library.Multiset_Order
HOL-Library: theory HOL-Library.Permutation
HOL-Library: theory HOL-Library.Char_ord
HOL-Library: theory HOL-Library.Code_Abstract_Nat
HOL-Library: theory HOL-Library.Code_Binary_Nat
HOL-Library: theory HOL-Library.Code_Target_Nat
HOL-Library: theory HOL-Library.Code_Prolog
HOL-Library: theory HOL-Library.Code_Target_Int
HOL-Library: theory HOL-Library.Code_Target_Numeral
HOL-Library: theory HOL-Library.Code_Test
HOL-Library: theory HOL-Library.Comparator
HOL-Library: theory HOL-Library.Conditional_Parametricity
HOL-Library: theory HOL-Library.Confluence
HOL-Library: theory HOL-Library.Confluent_Quotient
HOL-Library: theory HOL-Library.Dlist
HOL-Library: theory HOL-Library.Datatype_Records
HOL-Library: theory HOL-Library.Debug
HOL-Library: theory HOL-Library.Disjoint_Sets
HOL-Library: theory HOL-Library.Dual_Ordered_Lattice
HOL-Library: theory HOL-Library.Permutations
HOL-Library: theory HOL-Library.Fun_Lexorder
HOL-Library: theory HOL-Library.FuncSet
HOL-Library: theory HOL-Library.Equipollence
HOL-Library: theory HOL-Library.Function_Algebras
HOL-Library: theory HOL-Library.Function_Division
HOL-Library: theory HOL-Library.Groups_Big_Fun
HOL-Library: theory HOL-Library.IArray
HOL-Library: theory HOL-Library.Infinite_Set
HOL-Library: theory HOL-Library.Omega_Words_Fun
HOL-Library: theory HOL-Library.Ramsey
HOL-Library: theory HOL-Library.LaTeXsugar
HOL-Library: theory HOL-Library.Lattice_Constructions
HOL-Library: theory HOL-Library.Lattice_Syntax
HOL-Library: theory HOL-Library.Combine_PER
HOL-Library: theory HOL-Library.Complete_Partial_Order2
HOL-Library: theory HOL-Library.ListVector
HOL-Library: theory HOL-Library.List_Lexorder
HOL-Library: theory HOL-Library.Mapping
HOL-Library: theory HOL-Library.More_List
HOL-Library: theory HOL-Library.Poly_Mapping
HOL-Library: theory HOL-Library.AList_Mapping
HOL-Library: theory HOL-Library.Nat_Bijection
HOL-Library: theory HOL-Library.Stream
HOL-Library: theory HOL-Library.Old_Datatype
HOL-Library: theory HOL-Library.Old_Recdef
HOL-Library: theory HOL-Library.Open_State_Syntax
HOL-Library: theory HOL-Library.Option_ord
HOL-Library: theory HOL-Library.Parallel
HOL-Library: theory HOL-Library.Pattern_Aliases
HOL-Library: theory HOL-Library.Perm
HOL-Library: theory HOL-Library.Phantom_Type
HOL-Library: theory HOL-Library.Cardinality
HOL-Library: theory HOL-Library.Power_By_Squaring
HOL-Library: theory HOL-Library.Predicate_Compile_Alternative_Defs
HOL-Library: theory HOL-Library.Predicate_Compile_Quickcheck
HOL-Library: theory HOL-Library.Preorder
HOL-Library: theory HOL-Library.Product_Lexorder
HOL-Library: theory HOL-Library.Numeral_Type
HOL-Library: theory HOL-Library.Product_Plus
HOL-Library: theory HOL-Library.Product_Order
HOL-Library: theory HOL-Library.Finite_Lattice
HOL-Library: theory HOL-Library.Type_Length
HOL-Library: theory HOL-Library.Saturated
HOL-Library: theory HOL-Library.Quantified_Premise_Simproc
HOL-Library: theory HOL-Library.Quotient_Syntax
HOL-Library: theory HOL-Library.Quotient_Option
HOL-Library: theory HOL-Library.Quotient_Product
HOL-Library: theory HOL-Library.Quotient_Set
HOL-Library: theory HOL-Library.Quotient_List
HOL-Library: theory HOL-Library.Quotient_Sum
HOL-Library: theory HOL-Library.Quotient_Type
HOL-Library: theory HOL-Library.RBT_Impl
HOL-Library: theory HOL-Library.Realizers
HOL-Library: theory HOL-Library.Reflection
HOL-Library: theory HOL-Library.Refute
HOL-Library: theory HOL-Library.Rewrite
HOL-Library: theory HOL-Library.Set_Algebras
HOL-Library: theory HOL-Library.Sorting_Algorithms
HOL-Library: theory HOL-Library.Stirling
HOL-Library: theory HOL-Library.Sublist
HOL-Library: theory HOL-Library.Prefix_Order
HOL-Library: theory HOL-Library.Subseq_Order
HOL-Library: theory HOL-Library.Transitive_Closure_Table
HOL-Library: theory HOL-Library.Tree
HOL-Library: theory HOL-Library.Tree_Multiset
HOL-Library: theory HOL-Library.Uprod
HOL-Library: theory HOL-Library.While_Combinator
HOL-Library: theory HOL-Library.Bourbaki_Witt_Fixpoint
HOL-Library: theory HOL-Library.Z2
HOL-Library: theory HOL-Library.Countable
HOL-Library: theory HOL-Library.Countable_Set
HOL-Library: theory HOL-Library.Countable_Complete_Lattices
HOL-Library: theory HOL-Library.Countable_Set_Type
HOL-Library: theory HOL-Library.Set_Idioms
HOL-Library: theory HOL-Library.FSet
HOL-Library: theory HOL-Library.Finite_Map
HOL-Library: theory HOL-Library.RBT
HOL-Library: theory HOL-Library.RBT_Mapping
HOL-Library: theory HOL-Library.RBT_Set
HOL-Library: theory HOL-Library.BigO
HOL-Library: theory HOL-Library.Code_Real_Approx_By_Float
HOL-Library: theory HOL-Library.Diagonal_Subsequence
HOL-Library: theory HOL-Library.Discrete
HOL-Library: theory HOL-Library.Going_To_Filter
HOL-Library: theory HOL-Library.Indicator_Function
HOL-Library: theory HOL-Library.Landau_Symbols
HOL-Library: theory HOL-Library.Lattice_Algebras
HOL-Library: theory HOL-Library.Liminf_Limsup
HOL-Library: theory HOL-Library.Log_Nat
HOL-Library: theory HOL-Library.Lub_Glb
HOL-Library: theory HOL-Library.Multiset_Permutations
HOL-Library: theory HOL-Library.Nonpos_Ints
HOL-Library: theory HOL-Library.OptionalSugar
HOL-Library: theory HOL-Library.Order_Continuity
HOL-Library: theory HOL-Library.Extended_Nat
HOL-Library: theory HOL-Library.Interval
HOL-Library: theory HOL-Library.Float
HOL-Library: theory HOL-Library.Extended_Real
HOL-Library: theory HOL-Library.Interval_Float
HOL-Library: theory HOL-Library.Linear_Temporal_Logic_on_Streams
HOL-Library: theory HOL-Library.Periodic_Fun
HOL-Library: theory HOL-Library.Quadratic_Discriminant
HOL-Library: theory HOL-Library.Extended_Nonnegative_Real
HOL-Library: theory HOL-Library.Sum_of_Squares
HOL-Library: theory HOL-Library.Tree_Real
HOL-Library: theory HOL-Library.Library
HOL-Library FAILED
(see also /home/kuba/.isabelle/Isabelle2020/heaps/polyml-5.8.1_x86_64-linux/log/HOL-Library)
### Rule already declared as introduction (intro)
### \<lbrakk>closed ?S; open ?T\<rbrakk> \<Longrightarrow> closed (?S - ?T)
### Rule already declared as introduction (intro)
### closed ?S \<Longrightarrow> open (- ?S)
### Rule already declared as introduction (intro)
### open ?S \<Longrightarrow> closed (- ?S)
"- \<infinity>"
  :: "ereal"
"\<infinity>"
  :: "ereal"
"True"
  :: "bool"
"True"
  :: "bool"
"ereal (13 / 4)"
  :: "ereal"
### Rule already declared as introduction (intro)
### (\<And>x y. ?A x y \<Longrightarrow> ?B (?f x) (?g y)) \<Longrightarrow>
### rel_fun ?A ?B ?f ?g
### Ignoring duplicate rewrite rule:
### continuous_on ?s1 (\<lambda>x. ?c1) \<equiv> True
### Rule already declared as introduction (intro)
### (\<And>x y. ?A x y \<Longrightarrow> ?B (?f x) (?g y)) \<Longrightarrow>
### rel_fun ?A ?B ?f ?g
### Ignoring duplicate rewrite rule:
### continuous_on ?s1 (\<lambda>x. x) \<equiv> True
### Ignoring duplicate rewrite rule:
### continuous_on ?s1 (\<lambda>x. x) \<equiv> True
\<Sqinter> (Sup ` ?A) =
\<Squnion> (Inf ` {f ` ?A |f. \<forall>Y\<in>?A. f Y \<in> Y})
*** unknown Z3 proof rule "proof-bind"
*** At command "by" (line 297 of "~~/src/HOL/Library/Interval_Float.thy")
*** unknown Z3 proof rule "proof-bind"
*** At command "by" (line 2190 of "~~/src/HOL/Library/Float.thy")
*** unknown Z3 proof rule "proof-bind"
*** At command "by" (line 2119 of "~~/src/HOL/Library/Float.thy")
*** unknown Z3 proof rule "proof-bind"
*** At command "by" (line 1799 of "~~/src/HOL/Library/Poly_Mapping.thy")
*** unknown Z3 proof rule "proof-bind"
*** At command "by" (line 595 of "~~/src/HOL/Library/Ramsey.thy")
Unfinished session(s): HOL-Library
Return code: 2

Session build failed -- prover process remains inactive!
```

</details>

**Expected behavior**
The build should finish without an error.

**Additional context**
This does not happen on my Debian machine, with the distributed isabelle tarball. i believe the issue arises due to a mismatch of z3 versions - Isabelle ships with 4.4.0, but nixpkgs replaces it with 4.8.9.

**Notify maintainers**
@jwiegley 

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.65, NixOS, 21.03pre244932.84d74ae9c9c (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.7`
 - channels(kuba): `"nixpkgs-21.03pre244932.84d74ae9c9c"`
 - channels(root): `"nixos-21.03pre244932.84d74ae9c9c"`
 - nixpkgs: `/home/kuba/.nix-defexpr/channels/nixpkgs`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: isabelle
# a list of nixos modules affected by the problem
module:
```

Unfortunately, there's no `z3 4.4.0` in nixpkgs right now. I tried bringing back the definition from git history, but that yields the following error:

<details>

```
In file included from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/algorithm:60,
                 from ../src/util/hash.h:22,
                 from ../src/util/hashtable.h:26,
                 from ../src/util/str_hashtable.h:24,
                 from ../src/util/debug.cpp:24:
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/utility: In function 'constexpr typename std::tuple_element<_Int, std::pair<_Tp1, _Tp2> >::type&& std::get(std::pair<_Tp1, _Tp2>&&)':
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/utility:225:57: error: no matching function for call to 'move()'
  225 |     { return __pair_get<_Int>::__move_get(std::move(__in)); }
      |                                                         ^
In file included from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/nested_exception.h:40,
                 from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/exception:144,
                 from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/ios:39,
                 from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/ostream:38,
                 from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/iostream:39,
                 from ../src/util/debug.cpp:23:
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/move.h:99:5: note: candidate: 'template<class _Tp> constexpr typename std::remove_reference<_Tp>::type&& std::move(_Tp&&)'
   99 |     move(_Tp&& __t) noexcept
      |     ^~~~
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/move.h:99:5: note:   template argument deduction/substitution failed:
In file included from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/algorithm:60,
                 from ../src/util/hash.h:22,
                 from ../src/util/hashtable.h:26,
                 from ../src/util/str_hashtable.h:24,
                 from ../src/util/debug.cpp:24:
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/utility:225:57: note:   candidate expects 1 argument, 0 provided
  225 |     { return __pair_get<_Int>::__move_get(std::move(__in)); }
      |                                                         ^
In file included from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/char_traits.h:39,
                 from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/ios:40,
                 from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/ostream:38,
                 from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/iostream:39,
                 from ../src/util/debug.cpp:23:
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algobase.h:497:5: note: candidate: 'template<class _II, class _OI> _OI std::move(_II, _II, _OI)'
  497 |     move(_II __first, _II __last, _OI __result)
      |     ^~~~
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algobase.h:497:5: note:   template argument deduction/substitution failed:
In file included from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/algorithm:60,
                 from ../src/util/hash.h:22,
                 from ../src/util/hashtable.h:26,
                 from ../src/util/str_hashtable.h:24,
                 from ../src/util/debug.cpp:24:
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/utility:225:57: note:   candidate expects 3 arguments, 0 provided
  225 |     { return __pair_get<_Int>::__move_get(std::move(__in)); }
      |                                                         ^
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/utility: In function 'constexpr const typename std::tuple_element<_Int, std::pair<_Tp1, _Tp2> >::type&& std::get(const std::pair<_Tp1, _Tp2>&&)':
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/utility:235:63: error: no matching function for call to 'move()'
  235 |     { return __pair_get<_Int>::__const_move_get(std::move(__in)); }
      |                                                               ^
In file included from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/nested_exception.h:40,
                 from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/exception:144,
                 from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/ios:39,
                 from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/ostream:38,
                 from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/iostream:39,
                 from ../src/util/debug.cpp:23:
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/move.h:99:5: note: candidate: 'template<class _Tp> constexpr typename std::remove_reference<_Tp>::type&& std::move(_Tp&&)'
   99 |     move(_Tp&& __t) noexcept
      |     ^~~~
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/move.h:99:5: note:   template argument deduction/substitution failed:
In file included from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/algorithm:60,
                 from ../src/util/hash.h:22,
                 from ../src/util/hashtable.h:26,
                 from ../src/util/str_hashtable.h:24,
                 from ../src/util/debug.cpp:24:
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/utility:235:63: note:   candidate expects 1 argument, 0 provided
  235 |     { return __pair_get<_Int>::__const_move_get(std::move(__in)); }
      |                                                               ^
In file included from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/char_traits.h:39,
                 from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/ios:40,
                 from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/ostream:38,
                 from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/iostream:39,
                 from ../src/util/debug.cpp:23:
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algobase.h:497:5: note: candidate: 'template<class _II, class _OI> _OI std::move(_II, _II, _OI)'
  497 |     move(_II __first, _II __last, _OI __result)
      |     ^~~~
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algobase.h:497:5: note:   template argument deduction/substitution failed:
In file included from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/algorithm:60,
                 from ../src/util/hash.h:22,
                 from ../src/util/hashtable.h:26,
                 from ../src/util/str_hashtable.h:24,
                 from ../src/util/debug.cpp:24:
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/utility:235:63: note:   candidate expects 3 arguments, 0 provided
  235 |     { return __pair_get<_Int>::__const_move_get(std::move(__in)); }
      |                                                               ^
In file included from /nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/algorithm:62,
                 from ../src/util/hash.h:22,
                 from ../src/util/hashtable.h:26,
                 from ../src/util/str_hashtable.h:24,
                 from ../src/util/debug.cpp:24:
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algo.h: In function '_RandomAccessIterator std::__sample(_InputIterator, _InputIterator, std::input_iterator_tag, _RandomAccessIterator, std::random_access_iterator_tag, _Size, _UniformRandomBitGenerator&&)':
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algo.h:5762:21: error: expected ',' before '++' token
 5762 |    __out[__sample_sz++] = *__first;
      |                     ^~
      |                     ,
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algo.h:5762:21: error: expected identifier before '++' token
 5762 |    __out[__sample_sz++] = *__first;
      |                     ^~
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algo.h: In lambda function:
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algo.h:5762:25: error: expected '{' before '=' token
 5762 |    __out[__sample_sz++] = *__first;
      |                         ^
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algo.h: In lambda function:
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algo.h:5770:17: error: expected '{' before '=' token
 5770 |      __out[__k] = *__first;
      |                 ^
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algo.h: In function '_OutputIterator std::__sample(_ForwardIterator, _ForwardIterator, std::forward_iterator_tag, _OutputIterator, _Cat, _Size, _UniformRandomBitGenerator&&)':
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algo.h:5810:14: error: expected primary-expression before '=' token
 5810 |     *__out++ = *__first;
      |              ^
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algo.h:5821:14: error: expected primary-expression before '=' token
 5821 |     *__out++ = *__first;
      |              ^
/nix/store/h31cy7jm6g7cfqbhc5pm4rf9c53i3qfb-gcc-9.3.0/include/c++/9.3.0/bits/stl_algo.h:5834:15: error: expected primary-expression before '=' token
 5834 |      *__out++ = *__first;
      |               ^
../src/util/debug.cpp: In function 'void invoke_gdb()':
../src/util/debug.cpp:79:29: error: cannot convert 'std::basic_istream<char>' to 'bool' in initialization
   79 |         bool ok = (std::cin >> result);
      |                   ~~~~~~~~~~^~~~~~~~~~
      |                             |
      |                             std::basic_istream<char>
make: *** [Makefile:48: util/debug.o] Error 1
make: *** Waiting for unfinished jobs....
builder for '/nix/store/wdwm61nmqdr51dnibnjqj322wf91742i-z3-4.4.0.drv' failed with exit code 2
error: build of '/nix/store/wdwm61nmqdr51dnibnjqj322wf91742i-z3-4.4.0.drv' failed
```

</details>

I mentioned the issue in #isabelle on Freenode, and they are fairly certain that the z3 version is the issue. They also mentioned that, for this reason, they don't really like when distros package Isabelle...

--- a/pkgs/applications/science/logic/z3/4.4.0.nix
+++ b/pkgs/applications/science/logic/z3/4.4.0.nix
@@ -1,0 +1,41 @@
+{ stdenv, fetchFromGitHub, python }:
+
+stdenv.mkDerivation rec {
+  name = "z3-${version}";
+  version = "4.4.0";
+
+  src = fetchFromGitHub {
+    owner  = "Z3Prover";
+    repo   = "z3";
+    rev    = "7f6ef0b6c0813f2e9e8f993d45722c0e5b99e152";
+    sha256 = "1xllvq9fcj4cz34biq2a9dn2sj33bdgrzyzkj26hqw70wkzv1kzx";
+  };
+
+  buildInputs = [ python ];
+  enableParallelBuilding = true;
+
+  configurePhase = "python scripts/mk_make.py --prefix=$out && cd build";
+
+  # z3's install phase is stupid because it tries to calculate the
+  # python package store location itself, meaning it'll attempt to
+  # write files into the nix store, and fail.
+  soext = if stdenv.system == "x86_64-darwin" then ".dylib" else ".so";
+  installPhase = ''
+    mkdir -p $out/bin $out/lib/${python.libPrefix}/site-packages $out/include
+    cp ../src/api/z3*.h       $out/include
+    cp ../src/api/c++/z3*.h   $out/include
+    cp z3                     $out/bin
+    cp libz3${soext}          $out/lib
+    cp libz3${soext}          $out/lib/${python.libPrefix}/site-packages
+    cp z3*.pyc                $out/lib/${python.libPrefix}/site-packages
+    cp ../src/api/python/*.py $out/lib/${python.libPrefix}/site-packages
+  '';
+
+  meta = {
+    description = "A high-performance theorem prover and SMT solver";
+    homepage    = "http://github.com/Z3Prover/z3";
+    license     = stdenv.lib.licenses.mit;
+    platforms   = stdenv.lib.platforms.x86_64;
+    maintainers = with stdenv.lib.maintainers; [ thoughtpolice ttuegel ];
+  };
+}

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -26440,6 +26440,10 @@ in
   };
 
   z3 = callPackage ../applications/science/logic/z3 { python = python2; };
+  z3_4_4_0 = callPackage ../applications/science/logic/z3/4.4.0.nix {
+    python = python2;
+    stdenv = gcc49Stdenv;
+  };
   z3-tptp = callPackage ../applications/science/logic/z3/tptp.nix {};
 
   tlaplus = callPackage ../applications/science/logic/tlaplus {

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -26348,6 +26348,7 @@ in
     });
 
     java = openjdk11;
+    z3 = z3_4_4_0;
   };
 
   iprover = callPackage ../applications/science/logic/iprover { };

